libbeat/reader/multiline/pattern.go

```diff
@@ -0,0 +1,280 @@
+	}
+
+	matcher, err := matcherType(*config.Pattern)
+	if err != nil {
+		return nil, err
+	}
+
+	flushMatcher := config.FlushPattern
+
+	if config.Negate {
+		matcher = negatedMatcher(matcher)
+	}
+
+	maxLines := defaultMaxLines
+	if config.MaxLines != nil {
+		maxLines = *config.MaxLines
+	}
+
+	tout := defaultMultilineTimeout
+	if config.Timeout != nil {
+		tout = *config.Timeout
+		if tout < 0 {
+			return nil, fmt.Errorf("timeout %v must not be negative", config.Timeout)
+		}
+	}
+
+	if tout > 0 {
+		r = readfile.NewTimeoutReader(r, sigMultilineTimeout, tout)
+	}
+
+	pr := &patternReader{
+		reader:       r,
+		pred:         matcher,
+		flushMatcher: flushMatcher,
+		state:        (*patternReader).readFirst,
+		msgBuffer:    newMessageBuffer(maxBytes, maxLines, []byte(separator), config.SkipNewLine),
+		logger:       logp.NewLogger("reader_multiline"),
+	}
+	return pr, nil
+}
+
+// Next returns next multi-line event.
+func (pr *patternReader) Next() (reader.Message, error) {
+	return pr.state(pr)
+}
+
+func (pr *patternReader) readFirst() (reader.Message, error) {
+	for {
+		message, err := pr.reader.Next()
+		if err != nil {
+			// no lines buffered -> ignore timeout
+			if err == sigMultilineTimeout {
+				continue
+			}
+
+			pr.logger.Debug("Multiline event flushed because timeout reached.")
+
+			// pass error to caller (next layer) for handling
+			return message, err
+		}
+
+		if message.Bytes == 0 {
+			continue
+		}
+
+		// Start new multiline event
+		pr.msgBuffer.clear()
+		pr.msgBuffer.load(message)
+		pr.setState((*patternReader).readNext)
+		return pr.readNext()
+	}
+}
+
+func (pr *patternReader) readNext() (reader.Message, error) {
+	for {
+		message, err := pr.reader.Next()
+		if err != nil {
+			// handle multiline timeout signal
+			if err == sigMultilineTimeout {
+				// no lines buffered -> ignore timeout
+				if pr.msgBuffer.isEmpty() {
+					continue
+				}
+
+				pr.logger.Debug("Multiline event flushed because timeout reached.")
+
+				// return collected multiline event and
+				// empty buffer for new multiline event
+				msg := pr.msgBuffer.finalize()
+				pr.resetState()
+				return msg, nil
+			}
+
+			// handle error without any bytes returned from reader
+			if message.Bytes == 0 {
+				// no lines buffered -> return error
+				if pr.msgBuffer.isEmpty() {
+					return reader.Message{}, err
+				}
+
+				// lines buffered, return multiline and error on next read
+				msg := pr.msgBuffer.finalize()
+				pr.msgBuffer.setErr(err)
+				pr.setState((*patternReader).readFailed)
+				return msg, nil
+			}
+
+			// handle error with some content being returned by reader and
+			// line matching multiline criteria or no multiline started yet
+			if pr.msgBuffer.isEmptyMessage() || pr.pred(pr.msgBuffer.last, message.Content) {
+				pr.msgBuffer.addLine(message)
+
+				// return multiline and error on next read
+				msg := pr.msgBuffer.finalize()
+				pr.msgBuffer.setErr(err)
```

Comment:
`finalize` + `setErr` seem to be reoccuring patterns. How about adding a method like `failed(err error)` that executes the two?